tests/cpp/test_matmul_scheduler.cpp

```diff
@@ -2248,6 +2248,93 @@ TEST_F(MatmulSchedulerTest, StridedBatchEpilogueSingleBias) {
   }
 }
 
+// Test matmul with sizes that are not divisible by 8 and with misaligned inputs
+TEST_F(MatmulSchedulerTest, MisalignedVectorization) {
+  // TODO: parametrized test instead of nested loops (still use a loop over
+  // sizes and re-use FusionExecutorCache)
+  for (auto layout : kAllSupportedMmaLayout) {
+    for (bool downcast_output : {false, true}) {
+      // TODO: maybe use bias
+      auto fusion = std::make_unique<Fusion>();
+      FusionGuard fg(fusion.get());
+
+      auto tv0 = makeContigTensor(2, DataType::Half);
+      auto tv1 = makeContigTensor(2, DataType::Half);
+      fusion->addInput(tv0);
+      fusion->addInput(tv1);
+
+      tv0 = canonicalizeInputToBMNK(tv0, layout, MmaOperand::A);
+      tv1 = canonicalizeInputToBMNK(tv1, layout, MmaOperand::B);
+      auto tv2 = fusedMultiplySum(tv0, tv1, {-1});
+
+      if (downcast_output) {
+        tv2 = castOp(DataType::Half, tv2);
+      }
+
+      fusion->addOutput(tv2);
+
+      NVF_CHECK(
+          1 == ir_utils::getOpsOfType<MmaOp>(fusion.get()).size(),
+          "matmul fusion must have at least one MmaOp");
+      NVF_CHECK(
+          ir_utils::getOpsOfType<MmaOp>(fusion.get())
+              .front()
+              ->layout()
+              .has_value(),
+          "input layout has not be set for MmaOp");
+      NVF_CHECK(
+          MmaLayout::TN ==
+              ir_utils::getOpsOfType<MmaOp>(fusion.get())
+                  .front()
+                  ->layout()
+                  .value(),
+          "the MmaOp layout of Ampere MMA must be always TN");
+
+      const auto fusion_layout = mma_utils::getMmaLayout(fusion.get());
+      NVF_CHECK(
+          fusion_layout.isValid(),
+          "failed to get decide matmul layout through fusion definition");
+      NVF_CHECK(
+          fusion_layout.getData() == layout,
+          "mismatch between test layout (",
+          toString(layout),
+          ") and layout inferred from fusion definition (",
+          toString(fusion_layout.getData()),
+          ")");
+
+      FusionExecutorCache executor_cache(std::move(fusion));
+
+      for (const auto& [M, N, K, alignA, alignB, alignBias] :
+           std::vector<std::tuple<int, int, int, int, int, int>>{
+               {504, 136, 248, 16, 16, 16},
+               {505, 131, 249, 16, 16, 16},
```

Comment:
This case currently fails. We select the heuristic properly (no plugin, just `initCoreHeuristics`) I think:
```
===== Matmul Parameters ========

MMA macro: Ampere_16_16_16
DoubleBufferOptions:
  double_buffer_smem_write: true
  double_buffer_smem_read: true
  smem_double_buffer_stage: 2
SupportedVectorization:
  a: 1
  b: 1
  epilogue: 4
MatMulTileOptions: instruction tile [16, 16, 16], warp tile [64, 64, 32], CTA tile [256, 64, 32]
Rotate ldmatrix out of main loop: true
Async global mem load: false
Indexing mode: int32_t
Tile rastrization order: row-major
Grid swizzle factor: 1
Use shared memory epilogue: 0
Promote re-use of prologue shared memory: 1
Split-K factor: 1
```
However, execution fails with
>  what():  _result == CUDA_SUCCESS INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/executor_utils.cpp":1152, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. CUDA error: CUDA_ERROR_MISALIGNED_ADDRESS failed with error misaligned address
Generated kernel:
```c++
__global__ void nvfuser_matmul_f0_c1_r0_g0(Tensor<__half, 2, 2> T0, Tensor<__half, 2, 2> T1, Tensor<float, 2, 2> T5) {
  alignas(16) extern __shared__ char array[];
  const unsigned smem_offset = 0;
  nvfuser_index_t i0;
  i0 = ceilDiv(T0.logical_size[1LL], 32);
  nvfuser_index_t i1;
  i1 = ((nvfuser_index_t)threadIdx.z) % 2;
  nvfuser_index_t i2;
  i2 = ((nvfuser_index_t)threadIdx.x) / 8;
  nvfuser_index_t i3;
  i3 = (i2 + (4 * ((nvfuser_index_t)threadIdx.y))) + (4 * i1);
  nvfuser_index_t i4;
  i4 = ((nvfuser_index_t)threadIdx.z) / 2;
  nvfuser_index_t i5;
  i5 = ((nvfuser_index_t)threadIdx.x) % 8;
  nvfuser_index_t i6;
  i6 = (i5 + (64LL * i4)) + (64LL * (i3 / 8));
  nvfuser_index_t i7;
  i7 = i3 % 8;
  nvfuser_index_t i8;
  i8 = 32 * ((nvfuser_index_t)threadIdx.y);
  nvfuser_index_t i9;
  i9 = 32 * i1;
  nvfuser_index_t i10;
  i10 = T1.logical_size[1LL] * i4;
  nvfuser_index_t i11;
  i11 = 64LL * ((nvfuser_index_t)blockIdx.y);
  nvfuser_index_t i12;
  i12 = (((nvfuser_index_t)threadIdx.x) + i8) + i9;
  nvfuser_index_t i13;
  i13 = (i12 + i10) + i11;
  nvfuser_index_t i14;
  i14 = 2 * T1.logical_size[1LL];
  nvfuser_index_t i15;
  i15 = 32LL * ((nvfuser_index_t)threadIdx.y);
  nvfuser_index_t i16;
  i16 = 32LL * ((nvfuser_index_t)threadIdx.z);
  nvfuser_index_t i17;
  i17 = (((32LL * (i2 / 4)) + i5) + i15) + i16;
  nvfuser_index_t i18;
  i18 = i2 % 4;
  nvfuser_index_t i19;
  i19 = T0.logical_size[1LL] * ((nvfuser_index_t)threadIdx.y);
  nvfuser_index_t i20;
  i20 = T0.logical_size[1LL] * ((nvfuser_index_t)threadIdx.z);
  nvfuser_index_t i21;
  i21 = (256LL * T0.logical_size[1LL]) * ((nvfuser_index_t)blockIdx.x);
  nvfuser_index_t i22;
  i22 = ((((nvfuser_index_t)threadIdx.x) + i19) + i20) + i21;
  nvfuser_index_t i23;
  i23 = 4 * T0.logical_size[1LL];
  nvfuser_index_t i24;
  i24 = i2 / 2;
  nvfuser_index_t i25;
  i25 = i2 % 2;
  __half* T7 = reinterpret_cast<__half*>(array + smem_offset + 32768LL);
  __half* T6 = reinterpret_cast<__half*>(array + smem_offset + 0);
  unsigned i26;
  i26 = (toSmem(T6) + (512 * i25)) + (64 * i5);
  nvfuser_index_t i27;
  i27 = (i5 / 2) % 4;
  nvfuser_index_t i28;
  i28 = 4096 * ((nvfuser_index_t)threadIdx.z);
  unsigned i29;
  i29 = ((i26 + (64 * (i24 / 4))) + (16 * (i27 ^ (i24 % 4)))) + i28;
  unsigned i30;
  i30 = toSmem(T7);
  nvfuser_index_t i31;
  i31 = 128 * i5;
  nvfuser_index_t i32;
  i32 = 1024 * i25;
  nvfuser_index_t i33;
  i33 = 128 * ((nvfuser_index_t)threadIdx.y);
  unsigned i34;
  i34 = ((i30 + i31) + i32) + i33;
  nvfuser_index_t i35;
  i35 = i5 % 8;
  nvfuser_index_t i36;
  i36 = (((((32 * T1.logical_size[1LL]) + ((nvfuser_index_t)threadIdx.x)) + i8) + i9) + i10) + i11;
  nvfuser_index_t i37;
  i37 = 32LL * T1.logical_size[1LL];
  nvfuser_index_t i38;
  i38 = 32 + ((nvfuser_index_t)threadIdx.x);
  nvfuser_index_t i39;
  i39 = ((i38 + i19) + i20) + i21;
  nvfuser_index_t i40;
  i40 = 2LL + i24;
  unsigned i41;
  i41 = ((i26 + (64 * (i40 / 4))) + (16 * (i27 ^ (i40 % 4)))) + i28;
  unsigned i42;
  i42 = (((2048 + i30) + i31) + i32) + i33;
  nvfuser_index_t i43;
  i43 = ((nvfuser_index_t)threadIdx.x) / 4;
  nvfuser_index_t i44;
  i44 = 16LL * T1.logical_size[1LL];
  nvfuser_index_t i45;
  i45 = 2LL * (((nvfuser_index_t)threadIdx.x) % 4);
  nvfuser_index_t i46;
  i46 = 64LL * ((nvfuser_index_t)threadIdx.y);
  nvfuser_index_t i47;
  i47 = i43 / 8;
  nvfuser_index_t i48;
  i48 = i43 % 8;
  nvfuser_index_t i49;
  i49 = (((((i45 + (i44 * i47)) + (T1.logical_size[1LL] * i48)) + i46) + ((64LL * T1.logical_size[1LL]) * ((nvfuser_index_t)threadIdx.z))) + i11) + ((256LL * T1.logical_size[1LL]) * ((nvfuser_index_t)blockIdx.x));
  nvfuser_index_t i50;
  i50 = 8LL * T1.logical_size[1LL];
  bool b51;
  b51 = (i12 + i11) < T1.logical_size[1LL];
  nvfuser_index_t i52;
  i52 = (((-(64 * T0.logical_size[1LL])) + ((nvfuser_index_t)threadIdx.x)) + i15) + i16;
  bool b53;
  b53 = ((nvfuser_index_t)threadIdx.x) < T0.logical_size[1LL];
  nvfuser_index_t i54;
  i54 = -T0.logical_size[0LL];
  nvfuser_index_t i55;
  i55 = 256LL * ((nvfuser_index_t)blockIdx.x);
  nvfuser_index_t i56;
  i56 = ((i54 + ((nvfuser_index_t)threadIdx.y)) + ((nvfuser_index_t)threadIdx.z)) + i55;
  nvfuser_index_t i57;
  i57 = (32 - T0.logical_size[1LL]) + i4;
  nvfuser_index_t i58;
  i58 = (((1 - T1.logical_size[1LL]) + i45) + i46) + i11;
  nvfuser_index_t i59;
  i59 = (((i54 + (16LL * i47)) + i48) + (64LL * ((nvfuser_index_t)threadIdx.z))) + i55;
  Array<float, 128, 2> T8;
  #pragma unroll
  for(nvfuser_index_t i60 = 0; i60 < 4; ++i60) {
    nvfuser_index_t i61;
    i61 = 32 * i60;
    #pragma unroll
    for(nvfuser_index_t i62 = 0; i62 < 4; ++i62) {
      ((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i61 + (8 * i62))]))).set(0);
    }
  }
  #pragma unroll
  for(nvfuser_index_t i63 = 0; i63 < 16; ++i63) {
    T7[((i6 + (128 * i63)) + (8LL * (i7 ^ ((i4 + (2 * i63)) % 8))))] = 0;
  }
  #pragma unroll
  for(nvfuser_index_t i63 = 0; i63 < 16; ++i63) {
    if ((b51 && (i52 < (-(128LL * i63))))) {
      T7[((i6 + (128 * i63)) + (8LL * (i7 ^ ((i4 + (2 * i63)) % 8))))]
         = T1[(i13 + (i14 * i63))];
    }
  }
  #pragma unroll
  for(nvfuser_index_t i64 = 0; i64 < 64; ++i64) {
    T6[((i17 + (128 * i64)) + (8LL * (i18 ^ ((i4 + (2 * i64)) % 4))))] = 0;
  }
  #pragma unroll
  for(nvfuser_index_t i64 = 0; i64 < 64; ++i64) {
    if ((b53 && (i56 < (-(4 * i64))))) {
      T6[((i17 + (128 * i64)) + (8LL * (i18 ^ ((i4 + (2 * i64)) % 4))))]
         = T0[(i22 + (i23 * i64))];
    }
  }
  __syncthreads();
  Array<__half, 64, 8> T9;
  Array<__half, 64, 8> T3;
  #pragma unroll
  for(nvfuser_index_t i65 = 0; i65 < 4; ++i65) {
    asm volatile(
      "ldmatrix.sync.aligned.x4.m8n8.shared.b16 {%0, %1, %2, %3}, [%4];\n"
      :"=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(8LL * i65)]))[0]),
       "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(8LL * i65)]))[1]),
       "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(8LL * i65)]))[2]),
       "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(8LL * i65)]))[3])
      :"r"((uint32_t)((i29 + (1024 * i65))))
    );
  }
  #pragma unroll
  for(nvfuser_index_t i66 = 0; i66 < 4; ++i66) {
    nvfuser_index_t i67;
    i67 = i24 + (2LL * i66);
    asm volatile(
      "ldmatrix.sync.aligned.x4.trans.m8n8.shared.b16 {%0, %1, %2, %3}, [%4];\n"
      :"=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T3[(8LL * i66)]))[0]),
       "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T3[(8LL * i66)]))[1]),
       "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T3[(8LL * i66)]))[2]),
       "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T3[(8LL * i66)]))[3])
      :"r"((uint32_t)(((i34 + (128 * (i67 / 8))) + (16 * (i35 ^ (i67 % 8))))))
    );
  }
  #pragma unroll 1
  for(nvfuser_index_t i68 = 0; i68 < i0; ++i68) {
    nvfuser_index_t i69;
    i69 = (1 + i68) % 2;
    nvfuser_index_t i70;
    i70 = i6 + (2048 * i69);
    nvfuser_index_t i71;
    i71 = i36 + (i37 * i68);
    nvfuser_index_t i72;
    i72 = i17 + (8192 * i69);
    nvfuser_index_t i73;
    i73 = 32LL * i68;
    nvfuser_index_t i74;
    i74 = i39 + i73;
    nvfuser_index_t i75;
    i75 = i68 % 2;
    unsigned i76;
    i76 = i41 + (16384 * i75);
    unsigned i77;
    i77 = i42 + (4096 * i75);
    unsigned i78;
    i78 = i29 + (16384 * i69);
    unsigned i79;
    i79 = i34 + (4096 * i69);
    nvfuser_index_t i80;
    i80 = i57 + i73;
    bool b81;
    b81 = (i38 + i73) < T0.logical_size[1LL];
    #pragma unroll
    for(nvfuser_index_t i63 = 0; i63 < 16; ++i63) {
      T7[((i70 + (128 * i63)) + (8LL * (i7 ^ ((i4 + (2 * i63)) % 8))))] = 0;
    }
    #pragma unroll
    for(nvfuser_index_t i63 = 0; i63 < 16; ++i63) {
      nvfuser_index_t i82;
      i82 = 2 * i63;
      if ((b51 && (i80 < (-i82)))) {
        T7[((i70 + (128 * i63)) + (8LL * (i7 ^ ((i4 + i82) % 8))))]
           = T1[(i71 + (i14 * i63))];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i64 = 0; i64 < 64; ++i64) {
      T6[((i72 + (128 * i64)) + (8LL * (i18 ^ ((i4 + (2 * i64)) % 4))))] = 0;
    }
    #pragma unroll
    for(nvfuser_index_t i64 = 0; i64 < 64; ++i64) {
      if ((b81 && (i56 < (-(4 * i64))))) {
        T6[((i72 + (128 * i64)) + (8LL * (i18 ^ ((i4 + (2 * i64)) % 4))))]
           = T0[(i74 + (i23 * i64))];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i65 = 0; i65 < 4; ++i65) {
      asm volatile(
        "ldmatrix.sync.aligned.x4.m8n8.shared.b16 {%0, %1, %2, %3}, [%4];\n"
        :"=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(32 + (8LL * i65))]))[0]),
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(32 + (8LL * i65))]))[1]),
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(32 + (8LL * i65))]))[2]),
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(32 + (8LL * i65))]))[3])
        :"r"((uint32_t)((i76 + (1024 * i65))))
      );
    }
    __half T2[32];
    #pragma unroll
    for(nvfuser_index_t i83 = 0; i83 < 4; ++i83) {
      nvfuser_index_t i84;
      i84 = 8 * i83;
      #pragma unroll
      for(nvfuser_index_t i85 = 0; i85 < 2; ++i85) {
        nvfuser_index_t i86;
        i86 = i84 + (4LL * i85);
        #pragma unroll
        for(nvfuser_index_t i87 = 0; i87 < 2; ++i87) {
          nvfuser_index_t i88;
          i88 = i86 + (2LL * i87);
          #pragma unroll
          for(nvfuser_index_t i89 = 0; i89 < 2; ++i89) {
            T2[(i88 + i89)] = 0.000000000e+00f;
          }
        }
      }
    }
    #pragma unroll
    for(nvfuser_index_t i83 = 0; i83 < 4; ++i83) {
      nvfuser_index_t i90;
      i90 = 8 * i83;
      #pragma unroll
      for(nvfuser_index_t i85 = 0; i85 < 2; ++i85) {
        nvfuser_index_t i91;
        i91 = i90 + (4LL * i85);
        #pragma unroll
        for(nvfuser_index_t i87 = 0; i87 < 2; ++i87) {
          nvfuser_index_t i92;
          i92 = i91 + (2LL * i87);
          #pragma unroll
          for(nvfuser_index_t i89 = 0; i89 < 2; ++i89) {
            nvfuser_index_t i93;
            i93 = i92 + i89;
            T2[i93]
               = T9[i93];
          }
        }
      }
    }
    #pragma unroll
    for(nvfuser_index_t i66 = 0; i66 < 4; ++i66) {
      nvfuser_index_t i94;
      i94 = i24 + (2LL * i66);
      asm volatile(
        "ldmatrix.sync.aligned.x4.trans.m8n8.shared.b16 {%0, %1, %2, %3}, [%4];\n"
        :"=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T3[(32 + (8LL * i66))]))[0]),
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T3[(32 + (8LL * i66))]))[1]),
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T3[(32 + (8LL * i66))]))[2]),
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T3[(32 + (8LL * i66))]))[3])
        :"r"((uint32_t)(((i77 + (128 * (i94 / 8))) + (16 * (i35 ^ (i94 % 8))))))
      );
    }
    __half T4[32];
    #pragma unroll
    for(nvfuser_index_t i95 = 0; i95 < 4; ++i95) {
      nvfuser_index_t i96;
      i96 = 8 * i95;
      #pragma unroll
      for(nvfuser_index_t i97 = 0; i97 < 2; ++i97) {
        nvfuser_index_t i98;
        i98 = i96 + (4 * i97);
        #pragma unroll
        for(nvfuser_index_t i99 = 0; i99 < 2; ++i99) {
          nvfuser_index_t i100;
          i100 = i98 + (2LL * i99);
          #pragma unroll
          for(nvfuser_index_t i101 = 0; i101 < 2; ++i101) {
            T4[(i100 + i101)] = 0.000000000e+00f;
          }
        }
      }
    }
    #pragma unroll
    for(nvfuser_index_t i95 = 0; i95 < 4; ++i95) {
      nvfuser_index_t i102;
      i102 = 8 * i95;
      #pragma unroll
      for(nvfuser_index_t i97 = 0; i97 < 2; ++i97) {
        nvfuser_index_t i103;
        i103 = i102 + (4 * i97);
        #pragma unroll
        for(nvfuser_index_t i99 = 0; i99 < 2; ++i99) {
          nvfuser_index_t i104;
          i104 = i103 + (2LL * i99);
          #pragma unroll
          for(nvfuser_index_t i101 = 0; i101 < 2; ++i101) {
            nvfuser_index_t i105;
            i105 = i104 + i101;
            T4[i105]
               = T3[i105];
          }
        }
      }
    }
    #pragma unroll
    for(nvfuser_index_t i60 = 0; i60 < 4; ++i60) {
      nvfuser_index_t i106;
      i106 = 8 * i60;
      nvfuser_index_t i107;
      i107 = 32 * i60;
      #pragma unroll
      for(nvfuser_index_t i62 = 0; i62 < 4; ++i62) {
        nvfuser_index_t i108;
        i108 = 8 * i62;
        asm(
          "mma.sync.aligned.m16n8k16.row.col.f32.f16.f16.f32 {%0, %1, %2, %3}, {%4, %5, %6, %7}, {%8, %9}, {%10, %11, %12, %13};\n"
          :"=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[0LL][0]),
           "=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[0LL][1]),
           "=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[0LL][2]),
           "=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[0LL][3])
          :"r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i106]))[0]),
           "r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i106]))[1]),
           "r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i106]))[2]),
           "r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i106]))[3]),
           "r"((*reinterpret_cast<Array<Array<uint32_t, 2, 1>, 2, 1>*>(&((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T4[i108])))))[0LL][0]),
           "r"((*reinterpret_cast<Array<Array<uint32_t, 2, 1>, 2, 1>*>(&((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T4[i108])))))[0LL][1]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[0LL][0]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[0LL][1]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[0LL][2]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[0LL][3])
        );
        asm(
          "mma.sync.aligned.m16n8k16.row.col.f32.f16.f16.f32 {%0, %1, %2, %3}, {%4, %5, %6, %7}, {%8, %9}, {%10, %11, %12, %13};\n"
          :"=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[1LL][0]),
           "=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[1LL][1]),
           "=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[1LL][2]),
           "=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[1LL][3])
          :"r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i106]))[0]),
           "r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i106]))[1]),
           "r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i106]))[2]),
           "r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i106]))[3]),
           "r"((*reinterpret_cast<Array<Array<uint32_t, 2, 1>, 2, 1>*>(&((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T4[i108])))))[1LL][0]),
           "r"((*reinterpret_cast<Array<Array<uint32_t, 2, 1>, 2, 1>*>(&((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T4[i108])))))[1LL][1]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[1LL][0]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[1LL][1]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[1LL][2]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i107 + i108)])))))[1LL][3])
        );
      }
    }
    #pragma unroll
    for(nvfuser_index_t i83 = 0; i83 < 4; ++i83) {
      nvfuser_index_t i109;
      i109 = 8 * i83;
      #pragma unroll
      for(nvfuser_index_t i85 = 0; i85 < 2; ++i85) {
        nvfuser_index_t i110;
        i110 = i109 + (4LL * i85);
        #pragma unroll
        for(nvfuser_index_t i87 = 0; i87 < 2; ++i87) {
          nvfuser_index_t i111;
          i111 = i110 + (2LL * i87);
          #pragma unroll
          for(nvfuser_index_t i89 = 0; i89 < 2; ++i89) {
            T2[(i111 + i89)] = 0.000000000e+00f;
          }
        }
      }
    }
    #pragma unroll
    for(nvfuser_index_t i83 = 0; i83 < 4; ++i83) {
      nvfuser_index_t i112;
      i112 = 8 * i83;
      nvfuser_index_t i113;
      i113 = 32 + i112;
      #pragma unroll
      for(nvfuser_index_t i85 = 0; i85 < 2; ++i85) {
        nvfuser_index_t i114;
        i114 = 4LL * i85;
        nvfuser_index_t i115;
        i115 = i112 + i114;
        nvfuser_index_t i116;
        i116 = i113 + i114;
        #pragma unroll
        for(nvfuser_index_t i87 = 0; i87 < 2; ++i87) {
          nvfuser_index_t i117;
          i117 = 2LL * i87;
          nvfuser_index_t i118;
          i118 = i115 + i117;
          nvfuser_index_t i119;
          i119 = i116 + i117;
          #pragma unroll
          for(nvfuser_index_t i89 = 0; i89 < 2; ++i89) {
            T2[(i118 + i89)]
               = T9[(i119 + i89)];
          }
        }
      }
    }
    #pragma unroll
    for(nvfuser_index_t i95 = 0; i95 < 4; ++i95) {
      nvfuser_index_t i120;
      i120 = 8 * i95;
      #pragma unroll
      for(nvfuser_index_t i97 = 0; i97 < 2; ++i97) {
        nvfuser_index_t i121;
        i121 = i120 + (4 * i97);
        #pragma unroll
        for(nvfuser_index_t i99 = 0; i99 < 2; ++i99) {
          nvfuser_index_t i122;
          i122 = i121 + (2LL * i99);
          #pragma unroll
          for(nvfuser_index_t i101 = 0; i101 < 2; ++i101) {
            T4[(i122 + i101)] = 0.000000000e+00f;
          }
        }
      }
    }
    #pragma unroll
    for(nvfuser_index_t i95 = 0; i95 < 4; ++i95) {
      nvfuser_index_t i123;
      i123 = 8 * i95;
      nvfuser_index_t i124;
      i124 = 32 + i123;
      #pragma unroll
      for(nvfuser_index_t i97 = 0; i97 < 2; ++i97) {
        nvfuser_index_t i125;
        i125 = 4 * i97;
        nvfuser_index_t i126;
        i126 = i123 + i125;
        nvfuser_index_t i127;
        i127 = i124 + i125;
        #pragma unroll
        for(nvfuser_index_t i99 = 0; i99 < 2; ++i99) {
          nvfuser_index_t i128;
          i128 = 2LL * i99;
          nvfuser_index_t i129;
          i129 = i126 + i128;
          nvfuser_index_t i130;
          i130 = i127 + i128;
          #pragma unroll
          for(nvfuser_index_t i101 = 0; i101 < 2; ++i101) {
            T4[(i129 + i101)]
               = T3[(i130 + i101)];
          }
        }
      }
    }
    #pragma unroll
    for(nvfuser_index_t i60 = 0; i60 < 4; ++i60) {
      nvfuser_index_t i131;
      i131 = 8 * i60;
      nvfuser_index_t i132;
      i132 = 32 * i60;
      #pragma unroll
      for(nvfuser_index_t i62 = 0; i62 < 4; ++i62) {
        nvfuser_index_t i133;
        i133 = 8 * i62;
        asm(
          "mma.sync.aligned.m16n8k16.row.col.f32.f16.f16.f32 {%0, %1, %2, %3}, {%4, %5, %6, %7}, {%8, %9}, {%10, %11, %12, %13};\n"
          :"=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[0LL][0]),
           "=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[0LL][1]),
           "=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[0LL][2]),
           "=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[0LL][3])
          :"r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i131]))[0]),
           "r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i131]))[1]),
           "r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i131]))[2]),
           "r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i131]))[3]),
           "r"((*reinterpret_cast<Array<Array<uint32_t, 2, 1>, 2, 1>*>(&((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T4[i133])))))[0LL][0]),
           "r"((*reinterpret_cast<Array<Array<uint32_t, 2, 1>, 2, 1>*>(&((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T4[i133])))))[0LL][1]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[0LL][0]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[0LL][1]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[0LL][2]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[0LL][3])
        );
        asm(
          "mma.sync.aligned.m16n8k16.row.col.f32.f16.f16.f32 {%0, %1, %2, %3}, {%4, %5, %6, %7}, {%8, %9}, {%10, %11, %12, %13};\n"
          :"=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[1LL][0]),
           "=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[1LL][1]),
           "=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[1LL][2]),
           "=f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[1LL][3])
          :"r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i131]))[0]),
           "r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i131]))[1]),
           "r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i131]))[2]),
           "r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T2[i131]))[3]),
           "r"((*reinterpret_cast<Array<Array<uint32_t, 2, 1>, 2, 1>*>(&((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T4[i133])))))[1LL][0]),
           "r"((*reinterpret_cast<Array<Array<uint32_t, 2, 1>, 2, 1>*>(&((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T4[i133])))))[1LL][1]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[1LL][0]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[1LL][1]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[1LL][2]),
           "f"((*reinterpret_cast<Array<Array<float, 4, 1>, 2, 1>*>(&((*reinterpret_cast<Array<float, 8, 1>*>(&T8[(i132 + i133)])))))[1LL][3])
        );
      }
    }
    __syncthreads();
    #pragma unroll
    for(nvfuser_index_t i65 = 0; i65 < 4; ++i65) {
      asm volatile(
        "ldmatrix.sync.aligned.x4.m8n8.shared.b16 {%0, %1, %2, %3}, [%4];\n"
        :"=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(8LL * i65)]))[0]),
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(8LL * i65)]))[1]),
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(8LL * i65)]))[2]),
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(8LL * i65)]))[3])
        :"r"((uint32_t)((i78 + (1024 * i65))))
      );
    }
    #pragma unroll
    for(nvfuser_index_t i66 = 0; i66 < 4; ++i66) {
      nvfuser_index_t i134;
      i134 = i24 + (2LL * i66);
      asm volatile(
        "ldmatrix.sync.aligned.x4.trans.m8n8.shared.b16 {%0, %1, %2, %3}, [%4];\n"
        :"=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T3[(8LL * i66)]))[0]),
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T3[(8LL * i66)]))[1]),
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T3[(8LL * i66)]))[2]),
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T3[(8LL * i66)]))[3])
        :"r"((uint32_t)(((i79 + (128 * (i134 / 8))) + (16 * (i35 ^ (i134 % 8))))))
      );
    }
  }
  #pragma unroll
  for(nvfuser_index_t i135 = 0; i135 < 4; ++i135) {
    nvfuser_index_t i136;
    i136 = 32 * i135;
    nvfuser_index_t i137;
    i137 = i49 + (i44 * i135);
    nvfuser_index_t i138;
    i138 = -(16LL * i135);
    #pragma unroll
    for(nvfuser_index_t i139 = 0; i139 < 4; ++i139) {
      nvfuser_index_t i140;
      i140 = i136 + (8 * i139);
      nvfuser_index_t i141;
      i141 = 16LL * i139;
      nvfuser_index_t i142;
      i142 = i137 + i141;
      nvfuser_index_t i143;
      i143 = -i141;
      #pragma unroll
      for(nvfuser_index_t i144 = 0; i144 < 2; ++i144) {
        nvfuser_index_t i145;
        i145 = i140 + (4LL * i144);
        nvfuser_index_t i146;
        i146 = 8LL * i144;
        nvfuser_index_t i147;
        i147 = i142 + i146;
        bool b148;
        b148 = i58 < (i143 - i146);
        #pragma unroll
        for(nvfuser_index_t i149 = 0; i149 < 2; ++i149) {
          if ((b148 && (i59 < (i138 - (8LL * i149))))) {
            loadLocalToGlobal<float, /*vec_size=*/2, /*is_volatile=*/false>( &T5[(i147 + (i50 * i149))], &T8[(i145 + (2LL * i149))]);
          }
        }
      }
    }
  }
}
```

Reply:
Oh ok, the problem is in the output. Currently we are fully vectorizing the output, but that is not always possible depending on M and N.